test/modellib/ScanModel.cpp

```diff
@@ -0,0 +1,206 @@
+#include "test/modellib/ModelLib.hpp"
+
+#undef PRINT_TENSORS
+
+using namespace mlir;
+
+namespace onnx_mlir {
+namespace test {
+
+//
+// Functions to build onnx.Scan and onnx.Scan9 op for testing.
+//
+// Operation:
+//   onnx.Scan9
+// Attribute:
+//   Body: {onnx.Add}
+//   num_scan_inputs: 1
+//   scan_input_axis:
+//   scan_input_directions: []
+//   scan_output_axis:
+//   scan_output_directions: []
+// Inputs:
+//   initial_state_and_scan_inputs: V
+// Outputs:
+//   final_state_and_scan_outputs: V
+//
+// Example Parameter:
+//   B(=batch-size):1 x S(=sequence-length):3 x I(=inner-dim):2
+// Variables:
+//   initial_state_and_scan_outputs<BxI>
+//   x<BxSxI>(Inputs) : [[1, 3, 5], [2, 4, 6]]<1x3x2>
+//   y<BxI>(=last-state) : [9, 12]
+//   z<BxSxI>(scan-output) : [[1, 4, 9], [2, 6, 12]]
+// : [0, 0]<1x2>
+
+//
+// Similar to the onnx.Loop numerical test code, we use a model builder
+// based on MLIR-representation in order to define onnx.Scan's body part
+// in the simplest way.
+// 
+
+std::string testScanIdentityAdd = R"(
+module {
+  func @main_graph(%arg0: tensor<%Ixf32>, %arg1: tensor<%Sx%Ixf32>) ->
+                   (tensor<%Ixf32>, tensor<%Sx%Ixf32>) {
+    %1:2 = "onnx.Scan"(%arg0, %arg1) ({
+    ^bb0(%body_arg0: tensor<%Ixf32>, %body_arg1: tensor<%Ixf32>):
+      %2 = "onnx.Add"(%body_arg0, %body_arg1) :
+           (tensor<%Ixf32>, tensor<%Ixf32>) -> tensor<%Ixf32>
+      %3 = "onnx.Identity"(%2) : (tensor<%Ixf32>) -> tensor<%Ixf32>
+      "onnx.Return"(%2, %3) : (tensor<%Ixf32>, tensor<%Ixf32>) -> ()
+    }) {num_scan_inputs = 1 : si64} :
+        (tensor<%Ixf32>, tensor<%Sx%Ixf32>)
+        -> (tensor<%Ixf32>, tensor<%Sx%Ixf32>)
+    "func.return"(%1#0, %1#1) : (tensor<%Ixf32>, tensor<%Sx%Ixf32>) -> ()
+  }
+  "onnx.EntryPoint"() {func = @main_graph, numInputs = 2 : i32, numOutputs = 2 : i32, signature = "[    ]"} : () -> ()
+})";
+
+ScanLibBuilder::ScanLibBuilder(const std::string &modelName,
+    const int /*batch=*/B, const int /*seq=*/S, const int /*inner-dim=*/I)
+    : ModelLibBuilder(modelName), B(B), S(S), I(I) {}
+
+bool ScanLibBuilder::build() {
+  initialShape = {B, I};
+  xShape = {B, S, I};
+
+  moduleIR = std::regex_replace(
+      testScanIdentityAdd, std::regex("%B"), std::to_string(B));
+  moduleIR = std::regex_replace(moduleIR, std::regex("%S"), std::to_string(S));
+  moduleIR = std::regex_replace(moduleIR, std::regex("%I"), std::to_string(I));
+  return true;
+}
+
+bool ScanLibBuilder::compileAndLoad() {
+  OwningOpRef<ModuleOp> moduleOp =
+      mlir::parseSourceString<ModuleOp>(moduleIR, &ctx);
+  OwningOpRef<ModuleOp> module(std::move(moduleOp));
+  if (compileModule(module, ctx, sharedLibBaseName, onnx_mlir::EmitLib) != 0)
+    return false;
+  exec = new ExecutionSession(
+      onnx_mlir::getTargetFilename(sharedLibBaseName, onnx_mlir::EmitLib));
+  return exec != nullptr;
+}
+
+bool ScanLibBuilder::compileAndLoad(const onnx_mlir::CompilerOptionList &list) {
```

Comment:
Why redefine this one here. It appears to be identical to the superclass
```C++
bool ModelLibBuilder::compileAndLoad(
    const onnx_mlir::CompilerOptionList &list) {
  if (setCompilerOptions(list) != CompilerSuccess)
    return false;
  return compileAndLoad();
}
```

Reply:
@AlexandreEichenberger Removed "compileAndLoad(const onnx_mlir::CompilerOptionList &list)." Thank you.